java-checks/src/main/java/org/sonar/java/checks/MissingPathVariableAnnotationCheck.java

```diff
@@ -68,56 +68,62 @@ public void visitNode(Tree tree) {
-        return type.isSubtypeOf("java.util.Map") && stringToString;
-      });
-
-    if (containsMap) {
+  private void checkParameters(MethodTree method, String annotation, Set<String> modelAttributePathVariable) {
+    if (containsMap(method)) {
       /*
        * If any of the method parameters is a map, we assume all path variables are captured
        * and there is no mismatch with path variables in the request mapping.
        */
       return;
     }
 
-    Set<String> pathVariablesNames = method.parameters().stream()
-      .map(variable -> pathVariableName(variable))
+    Set<String> unusedPathVariables= findUnusedPathVariables(method, annotation, modelAttributePathVariable);
+    if(!unusedPathVariables.isEmpty()) {
+      reportIssue(
+        annotation(method, annotation),
+        "Bind path variable \"" + String.join("\", \"", unusedPathVariables) + "\" to a method parameter.");
+    }
+  }
+
+  private static Set<String> findUnusedPathVariables(MethodTree method, String annotation, Set<String> modelAttributePathVariable) {
+    Set<String> pathVariablesUsedInArguments = method.parameters().stream()
+      .map(MissingPathVariableAnnotationCheck::pathVariableName)
       .flatMap(Optional::stream)
       .collect(Collectors.toSet());
 
-    extractPathArgumentFromMappingAnnotations(method, annotation)
-      .map(path -> extractPathVariables(path))
-      .map(pathVariables -> {
-        pathVariables.removeAll(pathVariablesNames);
+    return extractPathArgumentFromMappingAnnotations(method, annotation)
+      .map(MissingPathVariableAnnotationCheck::extractPathVariables)
+      .flatMap(pathVariables -> {
+        pathVariables.removeAll(pathVariablesUsedInArguments);
         pathVariables.removeAll(modelAttributePathVariable);
-        return pathVariables;
+        return pathVariables.stream();
       })
-      .filter(pathVariables -> !pathVariables.isEmpty())
-      .forEach(pathVariables -> reportIssue(
-        annotation(method, annotation),
-        "Bind path variable \"" + String.join("\", \"", pathVariables) + "\" to a method parameter."));
+      .collect(Collectors.toSet());
+  }
+
+  private static boolean containsMap(MethodTree method) {
+    return method.parameters().stream()
+      .filter(parameter -> parameter.symbol().metadata().isAnnotatedWith(PATH_VARIABLE_ANNOTATION))
+      .anyMatch(parameter -> {
+        Type type = parameter.type().symbolType();
+        return type.isSubtypeOf("java.util.Map");
+      });
   }
 
   private static ExpressionTree annotation(MethodTree method, String name) {
     return method.modifiers().annotations().stream()
       .filter(annotation -> annotation.symbolType().is(name))
-      .findFirst()
-      // it will never be null because we are filtering on the annotation before.
-      .orElse(null);
+      .collect(Collectors.toList())
+      // it will never be empty because we are filtering on the annotation before.
+      .get(0);
```

Comment:
Should be changed into (as discussed in slack):

```
      .findFirst()
      // it will never be null because we are filtering on the annotation before.
      .orElseThrow();
```